node/src/main/scala/coop/rchain/node/proposescript/ProposeLeaderful.scala

```diff
@@ -0,0 +1,340 @@
+        .last
+        .map(_.getOrElse(false))
+
+    def waitFollowersGotLeaderBlock(blockHash: String): Task[Unit] =
+      for {
+        _ <- log.info(
+              s"Wait ${config.checkInterval} seconds and then check if followers got the block from leader."
+            )
+        _      <- wait(config.checkInterval)
+        result <- checkLeaderBlockInFollower(blockHash = blockHash)
+        _      <- waitFollowersGotLeaderBlock(blockHash).whenA(!result)
+      } yield ()
+
+    def proposeInFollowers =
+      Stream
+        .emits(config.followers)
+        .parEvalMapUnordered[Task, String](config.followers.length)(
+          v => v.deployAndProposeAsync.flatTap(blockHash => logValidatorProposedBlock(v, blockHash))
+        )
+        .compile
+        .toList
+
+    def checkFollowerBlocksInLeader(blockHashes: List[String]) =
+      Stream
+        .emits(blockHashes)
+        .parEvalMapUnordered[Task, Boolean](blockHashes.length)(
+          b =>
+            config.leader
+              .isContainBlock(b)
+              .flatTap { isContain =>
+                logValidatorGotBlock(config.leader, b)
+                  .whenA(isContain) >> logValidatorMissedBlock(config.leader, b).whenA(!isContain)
+              }
+        )
+        .forall(identity)
+        .compile
+        .last
+        .map(_.getOrElse(false))
+
+    def waitLeaderGotFollowerBlocks(blocks: List[String]): Task[Unit] =
+      for {
+        _ <- log.info(
+              s"Wait ${config.checkInterval} seconds and then check if leader got the block from followers."
+            )
+        _      <- wait(config.checkInterval)
+        result <- checkFollowerBlocksInLeader(blocks)
+        _      <- waitLeaderGotFollowerBlocks(blocks).whenA(!result)
+      } yield ()
+
+    val proposeRound = for {
+      _                    <- logWaitProposeInterval
+      _                    <- wait(config.proposeInterval)
+      _                    <- log.info(s"Going to propose in the leader ${config.leader.host}.")
+      proposeBlockInLeader <- proposeInLeader
+      _ <- log.info(
+            s"Proposed successfully in the leader with new block ${proposeBlockInLeader} " +
+              s"and then waiting the followers got the block."
+          )
+      _ <- waitFollowersGotLeaderBlock(proposeBlockInLeader)
+      _ <- logWaitProposeInterval
+      _ <- wait(config.proposeInterval)
+      _ <- log.info(
+            s"All followers got the newly proposed block ${proposeBlockInLeader} from leader."
+          )
+      proposeBlocksInFollowers <- proposeInFollowers
+      _ <- log.info(
+            s"Proposed successfully among all the followers. " +
+              s"Then waiting the leader to receive all the new blocks ${proposeBlocksInFollowers}."
+          )
+      _ <- waitLeaderGotFollowerBlocks(proposeBlocksInFollowers)
+      _ <- log.info(
+            "The leader successfully received all proposed blocks from followers."
+          )
+    } yield ()
```

Comment:
This code looks so nice that it deserves their own file and with abstracted F. Or keep it in this file and have new Main.